File: app/csharp-2/blazor/PuyoPuyoTDD.Tests/Models/GameTest.cs
// <copyright file="GameTest.cs" company="PuyoPuyoTDD">
// Copyright (c) PuyoPuyoTDD. All rights reserved.
// </copyright>

using PuyoPuyoTDD.Models;
using Xunit;

namespace PuyoPuyoTDD.Tests.Models;

/// <summary>
/// Game クラスのテスト.
/// </summary>
public class GameTest
{
    /// <summary>
    /// ゲームを初期化すると必要なコンポーネントが作成されるかテスト.
    /// </summary>
    [Fact]
    public void ゲームを初期化すると必要なコンポーネントが作成される()
    {
        // Arrange & Act
        var game = new Game();
        game.Initialize();

        // Assert
        Assert.NotNull(game.Config);
        Assert.NotNull(game.Stage);
        Assert.NotNull(game.Player);
        Assert.NotNull(game.Score);
    }

    /// <summary>
    /// ゲームを初期化するとゲームモードがStartになるかテスト.
    /// </summary>
    [Fact]
    public void ゲームを初期化するとゲームモードがStartになる()
    {
        // Arrange
        var game = new Game();

        // Act
        game.Initialize();

        // Assert
        Assert.Equal(GameMode.Start, game.Mode);
    }

    /// <summary>
    /// リスタートするとゲームモードがStartに戻るかテスト.
    /// </summary>
    [Fact]
    public void リスタートするとゲームモードがStartに戻る()
    {
        // Arrange
        var game = new Game();
        game.Initialize();

        // ゲームオーバー状態にする
        game.Stage.SetPuyo(2, 0, 1);
        game.Player.CreateNewPuyo();

        // Act
        game.Restart();

        // Assert
        Assert.Equal(GameMode.Start, game.Mode);
    }

    /// <summary>
    /// リスタートするとステージが初期化されるかテスト.
    /// </summary>
    [Fact]
    public void リスタートするとステージが初期化される()
    {
        // Arrange
        var game = new Game();
        game.Initialize();

        // ステージにぷよを配置
        game.Stage.SetPuyo(2, 10, 1);
        game.Stage.SetPuyo(3, 10, 2);

        // Act
        game.Restart();

        // Assert
        Assert.Equal(0, game.Stage.GetPuyo(2, 10));
        Assert.Equal(0, game.Stage.GetPuyo(3, 10));
    }

    /// <summary>
    /// リスタートするとスコアがリセットされるかテスト.
    /// </summary>
    [Fact]
    public void リスタートするとスコアがリセットされる()
    {
        // Arrange
        var game = new Game();
        game.Initialize();

        // スコアを加算
        game.Score.Add(1000);

        // Act
        game.Restart();

        // Assert
        Assert.Equal(0, game.Score.Value);
    }
}

File: app/kotlin/puyo-puyo-tdd/src/commonTest/kotlin/GameTest.kt
package com.example.puyopuyo

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

class GameTest {
    @Test
    fun ゲームを初期化すると必要なコンポーネントが作成される() {
        // Arrange
        val game = Game()

        // Act
        game.initialize()

        // Assert
        assertNotNull(game.config)
        assertNotNull(game.stage)
        assertNotNull(game.player)
        assertNotNull(game.score)
    }

    @Test
    fun ゲームを初期化するとゲームモードがStartになる() {
        // Arrange
        val game = Game()

        // Act
        game.initialize()

        // Assert
        assertEquals(GameMode.Start, game.mode)
    }

    @Test
    fun ゲームをリセットするとステージがクリアされる() {
        // Arrange
        val game = Game()
        game.initialize()
        game.player.createNewPuyo()
        // ステージにぷよを配置
        game.stage.setPuyo(0, 0, 1)
        game.stage.setPuyo(1, 1, 2)

        // Act
        game.reset()

        // Assert
        // ステージが空になっている
        for (y in 0 until game.config.stageHeight) {
            for (x in 0 until game.config.stageWidth) {
                assertEquals(0, game.stage.getPuyo(x, y))
            }
        }
    }

    @Test
    fun ゲームをリセットするとスコアがゼロになる() {
        // Arrange
        val game = Game()
        game.initialize()
        game.score.add(1000)

        // Act
        game.reset()

        // Assert
        assertEquals(0, game.score.value)
    }

    @Test
    fun ゲームをリセットすると新しいぷよが生成される() {
        // Arrange
        val game = Game()
        game.initialize()
        game.player.createNewPuyo()

        // Act
        game.reset()

        // Assert
        // 新しいぷよが生成されている（puyoType が 0 でない）
        assertNotNull(game.player.puyoType)
    }
}
